src/service/video-manager-impl.js

```diff
@@ -894,13 +945,235 @@ class VideoEntry {
+
+      const vidCenterX = dragCoord.position.x + minimizedRect.width / 2;
+      const vidCenterY = dragCoord.position.y + minimizedRect.height / 2;
+
+      if (vidCenterX > this.viewport_.getWidth()
+          || vidCenterX < 0
+          || vidCenterY > this.viewport_.getHeight()
+          || vidCenterY < 0) {
+        this.isDismissed_ = true;
+      }
+    } else {
+      dragCoord.position.x += dragCoord.velocity.x;
+      dragCoord.position.y += dragCoord.velocity.y;
+
+      dragCoord.velocity.x *= FRICTION_COEFF;
+      dragCoord.velocity.y *= FRICTION_COEFF;
+
+      if (this.isDismissed_) {
+        this.video.pause();
+        this.finishDocking_();
+        this.isDismissed_ = false;
+        return;
+      }
+    }
+
+    // Snap to corners
+    if (!this.isDragging_ && !this.isSnapping_
+        && Math.abs(dragCoord.velocity.x) <= STOP_THRESHOLD
+        && Math.abs(dragCoord.velocity.y) <= STOP_THRESHOLD) {
+      // X/Y Coordinates for each corner
+      const top = DOCK_MARGIN;
+      const left = DOCK_MARGIN;
+      const right = this.viewport_.getWidth()
+                    - minimizedRect.width
+                    - DOCK_MARGIN;
+      const bottom = this.viewport_.getHeight()
+                     - minimizedRect.height
+                     - DOCK_MARGIN;
+      // Determine corner and update this.dockPosition_
+      this.calcSnapCorner_(minimizedRect);
+      // Set coordinates based on corner
+      let newPosX = dragCoord.position.x, newPosY = dragCoord.position.y;
+      switch (this.dockPosition_) {
+        case DockPositions.BOTTOM_RIGHT:
+          newPosX = right;
+          newPosY = bottom;
+          break;
+        case DockPositions.TOP_RIGHT:
+          newPosX = right;
+          newPosY = top;
+          break;
+        case DockPositions.BOTTOM_LEFT:
+          newPosX = left;
+          newPosY = bottom;
+          break;
+        case DockPositions.TOP_LEFT:
+          newPosX = left;
+          newPosY = top;
+          break;
+      }
+      // Animate the snap transition
+      if (dragCoord.position.x != newPosX || dragCoord.position.y != newPosY) {
+        this.isSnapping_ = true;
+        // Snap to the calculated corner
+        this.animateSnap_(this.draggingMask_, newPosX, newPosY);
+        this.animateSnap_(this.internalElement_, newPosX, newPosY);
+      }
+    }
+
+    // Update the video's position
+    if (!this.isSnapping_) {
+      this.dragMove_(this.draggingMask_);
+      this.dragMove_(this.internalElement_);
+    }
+
+    // Re-run on every animation frame
+    this.vsync_.mutate(() => {
+      this.drag_();
+    });
+  }
+
+  /**
+   * Removes the draggable mask and ends dragging
+   * @private
+   */
+  finishDragging_() {
+    this.removeDraggingMask_();
+  }
+
+  /**
+   * Reads mouse coordinate and saves them to an internal variable
+   * @param {Event} e
+   * @param {boolean} updateDisplacement
+   * @private
+   */
+  mouse_(e, updateDisplacement = false) {
```

Comment:
related to comments above, we really need to separate measure and mutate and tied to events instead of running every rAF. How about something like:

1- video docked 
2- initilaizeDragdrop (in a measure block, cache whatever we need, install handlers, on `touchmove` call `dragtouchmove_`
3- `dragtouchmove_` does reads new coordinates from `e` and does a `mutate` (if we needed to reread anything then split it does both a `measure` and `mutate`)

We just wanna make sure we never measure during mutate and vice versa.